libirods_rule_engine_plugin-indexing.cpp

```diff
@@ -824,55 +829,192 @@ namespace
+					                       rule_text);
+					log_re::error(msg);
+					return ERROR(SYS_INVALID_INPUT_PARAM, std::move(msg));
+				}
+
+				rule_text = rule_text.substr(start, end - start);
+			}
+			// irule -F <script>
+			else if (const auto external_pos = rule_text.find("@external\n"); external_pos != std::string_view::npos) {
+				// If there are opening and closing curly braces following the "@external\n" prefix, then we
+				// can assume that the rule text most likely represents a JSON string.
+				if (const auto start = rule_text.find_first_of('{'); start != std::string_view::npos) {
+					const auto end = rule_text.rfind(" }");
+
+					if (end == std::string_view::npos) {
+						auto msg = fmt::format("Received malformed rule text. "
+						                       "Expected closing curly brace following rule text [{}].",
+						                       rule_text);
+						log_re::error(msg);
+						return ERROR(SYS_INVALID_INPUT_PARAM, std::move(msg));
+					}
+
+					rule_text = rule_text.substr(start, end - start);
+				}
+				// Otherwise, the rule text must represent something else. In this case, simply strip the
+				// "@external\n" prefix from the rule text and let the JSON parser throw an exception if the
+				// rule text cannot be parsed. This allows the REP to fail without causing the agent to crash.
+				else {
+					rule_text = rule_text.substr(external_pos + 10);
+				}
 			}
+
 			const auto rule_obj = json::parse(rule_text);
+#if 0
 			const std::string& rule_engine_instance_name = rule_obj["rule-engine-instance-name"];
 			// if the rule text does not have our instance name, fail
 			if (config->instance_name != rule_engine_instance_name) {
 				return ERROR(SYS_NOT_SUPPORTED, "instance name not found");
 			}
-#if 0
-			// catalog / index drift correction
-			if(irods::indexing::schedule::indexing ==
-			   rule_obj["rule-engine-operation"]) {
-				ruleExecInfo_t* rei{};
-				const auto err = _eff_hdlr("unsafe_ms_ctx", &rei);
-				if(!err.ok()) {
-					return err;
-				}
+#endif
+			ruleExecInfo_t* rei{};
+			const auto err = _eff_hdlr("unsafe_ms_ctx", &rei);
+			if (!err.ok()) {
+				return err;
+			}
 
-				const std::string& params = rule_obj["delay-parameters"];
+			const auto& op = rule_obj.at("rule-engine-operation").get_ref<const std::string&>();
+
+			if (!irods::is_privileged_client(*rei->rsComm)) {
+				const auto& user = rei->rsComm->clientUser;
+				const auto msg = fmt::format("{}: User [{}#{}] must be a rodsadmin to execute indexing operation [{}].",
+				                             __func__,
+				                             user.userName,
+				                             user.rodsZone,
+				                             op);
+				log_re::error(msg);
+				return ERROR(CAT_INSUFFICIENT_PRIVILEGE_LEVEL, std::move(msg));
+			}
 
-				json delay_obj;
-				delay_obj["rule-engine-operation"] = irods::indexing::policy::indexing;
+			if (irods::indexing::policy::object::index == op) {
```

Comment:
`irods::indexing::policy::object::index` expands to the string, `"irods_policy_indexing_object_index"`.

This is the rule name that must be used to do full text indexing of a single data object. It shares the same name as the rules which are fired as a result of triggering PEPs.

You can see the rest of the rule names here.
https://github.com/irods/irods_capability_indexing/blob/3f3529dea01c883f4e6c691108878fef0930a634/configuration.hpp#L189-L205

Are those the rule names we want admins to use or do we want to change them for manual execution contexts?

I can see value in the names staying as they are. It makes it easy for admins to know what happens because they will start to remember the names. However, admins may not be able to distinguish who/what invoked the rule.

All of that to say, perhaps the names should be changed to something like ...
- `indexing_index_data_object`
- `indexing_index_collection`
- `indexing_purge_data_object`

Note: This PR doesn't support invoking indexing rules via the NREP yet. It should be doable, but that may require changing the the rule names for correct behavior.

Reply:
don't feel too strongly either way yet.    consistency across our plugins is where i think i'd find the most value/good.

also noting that it's interesting our namespacing is not in the same order as the rule names...
'policy' and 'indexing'... switched places...

Reply:
I agree on the consistency thing.

As for the namespacing, that's not surprising to me. If the C++ code used `irods::policy`, the possibility of symbol collision rises since other plugins would likely follow suit and define things in the `irods::policy` namespace.

I don't know that "policy" is a term that's needed in the rule names since everything iRODS does is about policy.

Reply:
i think that convention started with policy composition... and hasn't really been codified/hardened yet.   TBD...